libavformat/rtcenc.c

```diff
@@ -0,0 +1,1689 @@
+
+end:
+    avio_context_free(&pb);
+    av_hmac_free(hmac);
+    return ret;
+}
+
+/**
+ * Open the UDP transport and complete the ICE handshake. Use fast retransmit to
+ * handle packet loss for the binding request.
+ *
+ * @return 0 if OK, AVERROR_xxx on error
+ */
+static int ice_handshake(AVFormatContext *s)
+{
+    int ret, size;
+    char url[256], buf[MAX_UDP_BUFFER_SIZE];
+    RTCContext *rtc = s->priv_data;
+    int fast_retries = rtc->ice_arq_max, timeout = rtc->ice_arq_timeout;
+
+    /* Build UDP URL and create the UDP context as transport. */
+    ff_url_join(url, sizeof(url), "udp", NULL, rtc->ice_host, rtc->ice_port, NULL);
+    ret = ffurl_alloc(&rtc->udp_uc, url, AVIO_FLAG_WRITE, &s->interrupt_callback);
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to open udp://%s:%d\n", rtc->ice_host, rtc->ice_port);
+        goto end;
+    }
+
+    av_opt_set(rtc->udp_uc->priv_data, "connect", "1", 0);
+    av_opt_set(rtc->udp_uc->priv_data, "fifo_size", "0", 0);
+    /* Set the max packet size to the buffer size. */
+    snprintf(buf, sizeof(buf), "%d", rtc->pkt_size);
+    av_opt_set(rtc->udp_uc->priv_data, "pkt_size", buf, 0);
+
+    ret = ffurl_connect(rtc->udp_uc, NULL);
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to connect udp://%s:%d\n", rtc->ice_host, rtc->ice_port);
+        goto end;
+    }
+
+    /* Make the socket non-blocking, set to READ and WRITE mode after connected */
+    ff_socket_nonblock(ffurl_get_file_handle(rtc->udp_uc), 1);
+    rtc->udp_uc->flags |= AVIO_FLAG_READ | AVIO_FLAG_NONBLOCK;
+
+    /* Build the STUN binding request. */
+    ret = ice_create_request(s, buf, sizeof(buf), &size);
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to create STUN binding request, size=%d\n", size);
+        goto end;
+    }
+
+    /* Fast retransmit the STUN binding request. */
+    do {
+        ret = ffurl_write(rtc->udp_uc, buf, size);
+        if (ret < 0) {
+            av_log(s, AV_LOG_ERROR, "Failed to send STUN binding request, size=%d\n", size);
+            goto end;
+        }
+
+        /* Wait so that the server can process the request and no need ARQ then. */
+#if ICE_PROCESSING_TIMEOUT > 0
+        av_usleep(ICE_PROCESSING_TIMEOUT * 10000);
+#endif
+
+        /* Read the STUN binding response. */
+        ret = ffurl_read(rtc->udp_uc, buf, sizeof(buf));
+        if (ret < 0) {
+            /* If max retries is 6 and start timeout is 21ms, the total timeout
+             * is about 21 + 42 + 84 + 168 + 336 + 672 = 1263ms. */
+            av_usleep(timeout * 1000);
+            timeout *= 2;
+
+            if (ret == AVERROR(EAGAIN) && fast_retries) {
+                fast_retries--;
+                continue;
+            }
+            av_log(s, AV_LOG_ERROR, "Failed to read STUN binding response, retries=%d\n", rtc->ice_arq_max);
+            goto end;
+        }
+    } while (ret < 0);
+
+    if (ret < 2 || buf[0] != 0x01 || buf[1] != 0x01) {
```

Comment:
need to process the stun binding request from the peer, and will try to make some patches.

`[rtc @ 0x15171bcb0] Invalid STUN binding response, size=104, type=0001`